### ISSUE TITLE
some problems about OP-TEE boot Linux on zc702

### ISSUE THREAD

**Issue by sarawell** [OP-TEE/optee_os#1478]
Hello folks,
      Recently,I'm trying to port optee to zynq-zc702. I've cloned the source codes from Github,and have compiled them.It was successful. Now, I'm trying to boot Linux as secondary through OP-TEE as primary.But I failed. It seems that OP-TEE does't boot Linux.I’m using the Xilinx version of Linux and U-boot that is configured for the zynq-zc702,
The console prints following messages. 
            Starting application at 0x3E000000 ...
            DEBUG:   TEE-CORE:add_phys_mem:257: CFG_SHMEM_START 4 0x3ff00000 size 0x00100000
            DEBUG:   TEE-CORE:add_phys_mem:257: CFG_TA_RAM_START 3 0x3e100000 size 0x01e00000
            DEBUG:   TEE-CORE:add_phys_mem:257: CFG_TEE_RAM_START 1 0x3e000000 size 0x00100000
            DEBUG:   TEE-CORE:add_phys_mem:257: SLCR_BASE 8 0xf8000000 size 0x00100000
            DEBUG:   TEE-CORE:add_phys_mem:257: PL310_BASE 8 0xf8f00000 size 0x00200000
            DEBUG:   TEE-CORE:add_phys_mem:257: GIC_BASE 8 0xf8f00000 size 0x00100000
            DEBUG:   TEE-CORE:add_phys_mem:269: Physical mem map overlaps 0xf8f00000
            DEBUG:   TEE-CORE:add_phys_mem:257: CONSOLE_UART_BASE 7 0xe0000000 size 0x00200000
            DEBUG:   TEE-CORE:add_va_space:294: type 9 size 0x00a00000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 1 0x3e000000..0x3e0fffff pa                                  0x3e000000..0x3e0fffff size 0x100000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 3 0x3e100000..0x3fefffff pa 0x3e100000..0x3fefffff size 0x1e00000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 4 0x3ff00000..0x3fffffff pa 0x3ff00000..0x3fffffff size 0x100000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 7 0x40000000..0x401fffff pa 0xe0000000..0xe01fffff size  0x200000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 8 0x40200000..0x402fffff pa 0xf8000000..0xf80fffff size 0x100000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 8 0x40300000..0x404fffff pa 0xf8f00000..0xf90fffff size 0x200000
            DEBUG:   TEE-CORE:init_mem_map:439: type va 9 0x40500000..0x40efffff pa 0x00000000..0x009fffff size 0xa00000
            INFO:    TEE-CORE: Initializing (Unknown #8 2017年 04月 11日 星期二 07:01:01 UTC arm)
            DEBUG:   [0x0] TEE-CORE:init_canaries:170: #Stack canaries for stack_tmp[0] with top at 0x3e05a5f8
            DEBUG:   [0x0] TEE-CORE:init_canaries:170: watch *0x3e05a5fc
            DEBUG:   [0x0] TEE-CORE:init_canaries:170: #Stack canaries for stack_tmp[1] with top at 0x3e05aaf8
            DEBUG:   [0x0] TEE-CORE:init_canaries:170: watch *0x3e05aafc
            DEBUG:   [0x0] TEE-CORE:init_canaries:171: #Stack canaries for stack_abt[0] with top at 0x3e05b338
            DEBUG:   [0x0] TEE-CORE:init_canaries:171: watch *0x3e05b33c
            DEBUG:   [0x0] TEE-CORE:init_canaries:171: #Stack canaries for stack_abt[1] with top at 0x3e05bb78
           DEBUG:   [0x0] TEE-CORE:init_canaries:171: watch *0x3e05bb7c
           DEBUG:   [0x0] TEE-CORE:init_canaries:173: #Stack canaries for stack_thread[0] with top at 0x3e05dbb8
          DEBUG:   [0x0] TEE-CORE:init_canaries:173: watch *0x3e05dbbc
          DEBUG:   [0x0] TEE-CORE:init_canaries:173: #Stack canaries for stack_thread[1] with top at 0x3e05fbf8
          DEBUG:   [0x0] TEE-CORE:init_canaries:173: watch *0x3e05fbfc
          INFO:    TEE-CORE: Initialized
          DEBUG:   [0x0] TEE-CORE:init_primary_helper:633: Primary CPU switching to normal world boot_

Is there anybody who can help?

**Comment by dnovick** [comment]
Greetings,

So, I have been working on a very similar issue for quite some time, and am currently stuck. So far, nobody has been able to help with, including Xilinx support. My issue manifests itself similarly  - i.e. successful OP-TEE boot, but failing Linux boot. Now, I'm running the stack on a ZedBoard, and am using a Xilinx Petalinux build, however I suspect things are sufficiently the same to conclude that we are fighting similar issues. I have spent countless hours narrowing down the exact place where the problem occurs, and have debugged until my eyes have fallen out. Way back when this all started, I found that a particular errata fix in the Linux startup was causing a problem (the particular escapes me at the moment). Once I commented it out, the problem went away, but something similar now manifests itself when Linux tries to initialize the ARM PLL. As I said above, nobody on any forum has been able to help, though I'm still trying to rule out a couple of root causes with Xilinx. I could give more details, but it would break my finger. It would be better to talk about it.

Let me suggest the following. Can I get your email info somehow? I'd be happy to chat with you to walk through what I've already done, and compare that to what you've tried. That way we can rule out those things I've already worked through...maybe accelerate your effort....and possible address the issue to both our satisfaction once and for all. 

Just so you know, I was supposed to do a demo to some senior VPs at my company almost 2 months ago on this platform, and have the entire HW/SW stack ready to go. I'm just blocked on this issue. Thankfully, they have been quite supportive. Nevertheless, I'm really eager to get this fixed. If as part of my efforts, I can help somebody else avoid the agony I've been enduring, I'm more than happy to.

BTW, I'm also targeting an Altera platform for the same demo.

Let me know what you think. As I'm sure you know, putting heads together can really pay off.

Thanks,
Dave

**Comment by jenswi-linaro** [comment]
Since we're talking about a Cortex-A9 platform there's the external cache (PL0310?) to consider. plat-zynq7k uses `CFG_PL310_LOCKED` to emulate a disabled external cache when Linux is booting. It could be worth double checking that you're doing something similar on your platform.

**Comment by dnovick** [comment]
I haven't checked the Petalinux code yet, but quick question...Is this cache config "trick" required in both scenarios, i.e. when Linux is boot in secure mode as well as non-secure mode...or does it not matter? I only ask as it seems most Linux work on these platforms focuses only on running Linux as the sole OS, in secure mode. 

I'll check to see how the Petalinux plat-zynq7k config handles this. I would expect that if this is a known requirement with A9 based platforms, then the Petalinux config for this platform would handle it properly.

Dave

**Comment by jenswi-linaro** [comment]
No, if Linux is booting in secure mode then it has full control over the external cache controller, just as in the pre-TrustZone platforms.
Linux requires all caches to be turned off when initially entered.

**Comment by dnovick** [comment]
OK...Thanks. So you say plat-zynq7k uses this setting. Where specifically is this setting used in the source and where is it configured? At the moment, I'm working with a petalinux build for a zed board which of course uses plat-zynq7k. However, I am unable to find this configuration setting anywhere in the petalinux distribution. I have grepped the entire petalinux subdir (all text files), and cannot find this setting anywhere. Before concluding that it isn't there for petalinux...and going looking for something analogous or concluding that it's missing, I wanted to check to make sure I understood your comment correctly, and am looking in the right places.

Thanks,.

**Comment by sarawell** [comment]
@dnovick 
Hi Dave
    I am sorry for the late reply. My email is 664943056@qq.com
    Aimed at running OP-TEE and Linux on the ZC702 platform, I have tried a lot and booted op-tee successfully. But I encountered some problems when using OP-TEE to boot Linux. By inserting debug information in the source code and tracing the running program, I have found some possible reasons that I can not confirm now. I notice that we are doing the similar project because of your issues (Problems booting Linux on Zedboard after OP-TEE boots), using different platforms though.
    I am glad to discuss this issue with you. 
    Looking forward to your reply.
    Best Regards!

**Comment by jbech-optee-adm** [comment]
We are closing this issue/question, why?
 - We believe the question has been answered.
 - Quite often when we have answered the questions, the issuer forget to close the issue which results in many open issues for us which makes it hard to keep track of the ones actually waiting for an answer.
 - In some cases we ask follow-up questions, but do not hear back from the one creating the issue.

If you however feel that you have additional questions or still thinks this is an issue, then please feel free to re-open the issue again. When closing issues we will only show this entire message once.

// OP-TEE admins

**Comment by akingscote** [comment]
@sarawell Did you get this working? im having the exact same problem

**Comment by duclongpl** [comment]
@sarawell : I had the same problem like you before. And I'm still stuck about this issue.

**Comment by akingscote** [comment]
@yanyan-wrs @coolfengyu Any ideas on how to resolve this?

**Comment by coolfengyu** [comment]
I have not tested Linux yet. Please check whether the Linux code accessed SLCR, PL310 and others modules which only can be accessed in secure world.
They only can be accessed or configured by SMC call.

**Comment by eloysu520** [comment]
I have the same problem here and don't have solution yet. But my problem seems it is not a problem of the configuration of Linux. After analyze of the **generic_entry_a32.S**, in the terminal there should be another output such as "Secondary CPU Switching to normal world boot", but it is missing. Then i added some debug codes into **generic_entry_a32.S** to find out, where is it exactly been stopped. 

- In **LOCAL_FUNC reset_primary**:
`/* release secondary boot cores and sync with them */ `	
` cpu_is_ready` 	
`bl debug`
`flush_cpu_semaphores`
`bl debug`
`wait_secondary`
`bl debug`
the program stops after execution of _flush_cpu_semaphores_. It looks like it is waiting for reset of the secondary CPU, and there is something wrong with the secondary CPU. 

- Then in **LOCAL_FUNC reset_secondary**:
`bl	core_init_mmu_regs`
`bl debug`
`bl	cpu_mmu_enable`
`bl debug`
_core_init_mmu_regs_ is executed, but the program freezes during the execution of _cpu_mmu_enable_. 

Any idea how to solve this?

**Comment by jenswi-linaro** [comment]
This issue is already answered and closed. Please open a new issue, feel free to refer to this issue if you think it helps in describing your problem.
